### PR TITLE
brew.rb: tell more people to not report issues.

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
- We should tell people to not report issues if they are running in an unsupported configuration.
- We should tell people to run `brew update` before reporting issues if they have `HOMEBREW_NO_AUTO_UPDATE` set.
- We should tell people to not report issues in more types of exceptions.
- Warn people in `HOMEBREW_NO_AUTO_UPDATE` documentation.
- DRY up `brew.rb` exception handling.

### CODE REVIEW

**Review by SMillerDev — APPROVED** [review]

File: Library/Homebrew/brew.rb [hunk 1]
```diff
@@ -148,7 +148,11 @@
   Utils::Analytics.report_build_error(e)
   e.dump(verbose: args&.verbose?)
 
-  if e.formula.head? || e.formula.deprecated? || e.formula.disabled?
+  if Homebrew::EnvConfig.no_auto_update?
+    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```
Comment (colindean) [inline Library/Homebrew/brew.rb:152]:
```suggestion
    $stderr.puts "#{Tty.bold}You have disabled automatic updates. Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```

File: Library/Homebrew/brew.rb [hunk 2]
```diff
@@ -170,6 +174,12 @@
 
   exit 1
 rescue RuntimeError, SystemCallError => e
+  if Homebrew::EnvConfig.no_auto_update?
+    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```
Comment (colindean) [inline Library/Homebrew/brew.rb:178]:
```suggestion
    $stderr.puts "#{Tty.bold}You have disabled automatic updates. Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```

**Review by razvanazamfirei — APPROVED** [review]

File: Library/Homebrew/brew.rb [hunk 3]
```diff
@@ -178,6 +188,13 @@
   exit 1
 rescue MethodDeprecatedError => e
   onoe e
+
+  if Homebrew::EnvConfig.no_auto_update?
+    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```
Comment (colindean) [inline Library/Homebrew/brew.rb:193]: With this same error message thrice, it's probably appropriate to put it in a variable. 

```suggestion
    $stderr.puts "#{Tty.bold}You have disabled automatic updates. Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```

File: Library/Homebrew/brew.rb [hunk 4]
```diff
@@ -186,15 +203,17 @@
...
-  if internal_cmd && !OS.unsupported_configuration?
-    if Homebrew::EnvConfig.no_auto_update?
-      $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
-    else
-      $stderr.puts "#{Tty.bold}Please report this issue:#{Tty.reset}"
-      $stderr.puts "  #{Formatter.url(OS::ISSUES_URL)}"
-    end
+
+  if Homebrew::EnvConfig.no_auto_update?
+    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```
Comment (colindean) [inline Library/Homebrew/brew.rb:208]:
```suggestion
    $stderr.puts "#{Tty.bold}You have disabled automatic updates. Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```

**Review by colindean — COMMENTED** [review]
Adding 

> You have disabled automatic updates.

to the error message clarifies why the error occurred. 

It would be kind to include how to reenable them, too, since that's the preferred path.

**Review by p-linnane — APPROVED** [review]

**Review by ZhongRuoyu — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
I've revamped this a bit to DRY up the repeated `brew.rb` logic.

**Review by Bo98 — APPROVED** [review]
Hard to tell from the diff but I think it's ok

File: docs/Manpage.md [hunk 5]
```diff
@@ -3740,8 +3740,10 @@ command execution e.g. `$(cat file)`.
 `HOMEBREW_NO_AUTO_UPDATE`
 
 : If set, do not automatically update before running some commands, e.g. `brew
-  install`, `brew upgrade` and `brew tap`. Alternatively, run this less often by
-  setting `HOMEBREW_AUTO_UPDATE_SECS` to a value higher than the default.
+  install`, `brew upgrade` and `brew tap`. Preferably, run this less often by
+  setting `HOMEBREW_AUTO_UPDATE_SECS` to a value higher than the default. Note
+  that setting and e.g. tapping new taps may result in a broken configuration.
```
Comment (EricFromCanada) [inline docs/Manpage.md:3745]:
```suggestion
  that this setting and e.g. tapping new taps may result in a broken configuration.
```

File: Library/Homebrew/brew.rb [hunk 6]
```diff
@@ -166,35 +168,35 @@
...
-    else
-      $stderr.puts "#{Tty.bold}Please report this issue:#{Tty.reset}"
-      $stderr.puts "  #{Formatter.url(OS::ISSUES_URL)}"
-    end
+
+  if OS.unsupported_configuration?
+    $stderr.puts "#{Tty.bold}Do not report this issue: you are running in an unsupported configuration.#{Tty.reset}"
+  elsif Homebrew::EnvConfig.no_auto_update?
+    $stderr.puts "#{Tty.bold}You have disabled automatic updates.#{Tty.reset}"
+    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again.#{Tty.reset}"
```
Comment (colindean) [inline Library/Homebrew/brew.rb:191]: **suggestion (non-blocking):** 
```suggestion
    $stderr.puts "#{Tty.bold}You have disabled automatic updates.#{Tty.reset}"
    $stderr.puts "#{Tty.bold}Do not report this issue until you've run `brew update` and tried again or unset HOMEBREW_NO_AUTO_UPDATE.#{Tty.reset}"
```

Soft pitch on the instruction to clear the envvar…

**Review by colindean — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks all!

**Review by ZhongRuoyu — APPROVED** [review]

File: Library/Homebrew/brew.rb [hunk 7]
```diff
@@ -166,35 +168,35 @@
...
+  elsif (issues_url = (method_deprecated_error && e.issues_url) || Utils::Backtrace.tap_error_url(e))
+    $stderr.puts "If reporting this issue please do so at (not Homebrew/brew or Homebrew/homebrew-core):"
+    $stderr.puts "  #{Formatter.url(issues_url)}"
+  elsif internal_cmd
+    $stderr.puts "#{Tty.bold}Please report this issue:#{Tty.reset}"
+    $stderr.puts "  #{Formatter.url(OS::ISSUES_URL)}"
```
Comment (carlocab) [inline Library/Homebrew/brew.rb:197]: This is probably a pre-existing issue, but: not sure these categories should be treated as mutually exclusive (which is kinda what happens as a consequence of the if-elsif-else structure).

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.rb:197]: @carlocab which ones do you think should not be exclusive? The first two are broadly "don't report this", the latter two are "here's where to report it" (in different places).

I could see it being worth actually checking when the last `brew update` was done and, if e.g. today, then we don't print the auto-update message?

Comment (cho-m) [inline Library/Homebrew/brew.rb:197]: It looks like refactor is causing this message to be output more often? Like on a non-existent formula, e.g.
```console
❯ brew uninstall test
Error: No available formula with the name "test". Did you mean testssl?
Please report this issue:
  https://docs.brew.sh/Troubleshooting

❯ git checkout 4.2.16
Note: switching to '4.2.16'.
...

❯ brew uninstall test
Error: No available formula with the name "test". Did you mean testssl?
```

In my opinion, it seems a little odd to tell users to report this.

Comment (cho-m) [reply Library/Homebrew/brew.rb:197]: Regression is probably from no longer just exiting on `RuntimeError`. We do classify a bunch of exceptions as `RuntimeError` (in this case, `FormulaUnavailableError < FormulaOrCaskUnavailableError < RuntimeError`).

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.rb:197]: @cho-m Thanks, will take a look.

**Pull request merged**